Create 'immutable artefacts' pattern (#201)
HO-CTO/engineering-guidance-and-standards: docs/patterns/immutable-artefacts.md

---
layout: sub-navigation
order: 1
title: Immutable artefacts
date: git Last Modified
tags:
- Build, release and deploy

> - Deployment

Reviewer: Does it need a patterns tag?

Author: Patterns tag is automatically applied during 11ty process, by virtue of being in the patterns subdirectory 🙂

- Artefacts
---

{% import "macros/tags.njk" as pageTag %}

Last updated: {{ page.date | postDate }}
{{ pageTag.tags(tags)  }}

> An artefact is the deployable unit of code, for instance a Docker image, serverless package, or virtual machine image. You should build artefacts once, that you never modify, and use them across all environments. In other words, if you need to make changes, a new artefact should be created. This helps with assurance and debugging incidents.

Reviewer: ```suggestion
An artefact is the deployable unit of code, for instance a Docker image, serverless package, or virtual machine image. You should build artefacts once, afterwards and you should never modify them. If you need to make changes, a new artefact should be created. This helps with assurance and debugging incidents. 

An artefact should then be promoted between environments, this enables you to know what is running as the same arefact and code is used across different environments.
```
Think this reads a little clunky, how's this?

Author: Have rewritten this entirely to be clearer on what an immutable artefact is

> With proper versioning of your artefacts, you will be able to tie your environments back to a specific commit in your code-base, which is useful for investigation of incidents. Without immutable artefacts you may have hidden state and/or subtle differences between environments, that make it very hard to understand the behaviour your are observing and undermine your confidence that something will continue to work when moved to another environment. You will also find it hard to recover from a future disaster, as you will need to rebuild your artefact first, and your build pipeline may be broken.

Reviewer: ```suggestion
With proper versioning of your artefacts, you will be able to tie your environments back to a specific commit in your code-base, which is useful for investigation of incidents. Without immutable artefacts you may have hidden state and/or subtle differences between environments, that make it very hard to understand the behaviour you are observing and undermine your confidence that something will continue to work when moved to another environment. You will also find it hard to recover from a future disaster, as you will need to rebuild your artefact first, and your build pipeline may be broken.
```

---

## Solution

> Immutable artefacts help us to know what we are running. They aid us in accurately reproducing environments and so help with:

Reviewer: ```suggestion
Immutable artefacts help us to know what we are running. They aid us in accurately reproducing environments, this helps with:
```

- debugging incidents
- assurance over promotions
- roll-back of failed changes
- disaster recovery (consider rebuilding an environment from scratch)

You should:

- Produce an immutable artefact early in your build pipeline and store it with a unique version in a repository. This should be automated where possible. For example:
```
${semver}-${build-number}-${commit-hash-fragment}
```
- Promote this artefact through each assurance step and link the results to the artefact to provide an audit record.

> - Keep configuration out of the artefact. This should be provided using environment variables or similar external methods. See also Store Infrastructure and Configuration as Code.

Reviewer: Can this be a link to the standard for this?

Author: Yes, can update this as soon as the infrastructure as code standard is merged

Author: Now that the infrastructure and configuration as code standard is focussed exclusively on infrastructure as code this link is not relevant any more

Immutable artefacts are artefacts that are not modified after they are built. They enable idempotent and repeatable deployment. They help us to know what we are running, and they aid us in accurately and consistently reproducing environments. With proper versioning of your artefacts, you will be able to tie your environments back to a specific commit in your code-base. This helps with:

- Investigating incidents and debugging
- Assurance when promoting code
- Roll-back of failed changes
- Disaster recovery
  
---

## Solution

You should follow a '[build once, deploy many](https://www.mikemcgarr.com/blog/build-once-deploy-many.html)' approach. Specifically, you should:

- When you create a new version of your application or service (typically, in response to merging some new code into your main branch), build your application once and package it into an artefact along with its dependencies

> - Store artefacts in a suitable artefact repository with a unique version identifier. Versioning should follow a common pattern, and be automated where possible. For example:
> ```
> ${semver}-${build-number}-${commit-hash-fragment}
> ```

Reviewer 2: ```suggestion
- Store artefacts in a suitable artefact repository with a unique version identifier. Versioning should follow a common pattern, and be automated where possible, for example using [Semantic Versioning](https://semver.org/) or simply the build number.
```

I think the example is a bit confusing. Suggest referring to SemVer; it covers things like pre-release builds.
